notebooks/plot_attention.ipynb

```diff
@@ -1,1115 +1,1110 @@
+                "from maze_transformer.training.tokenizer import maze_to_tokens, SPECIAL_TOKENS\n",
+                "from maze_transformer.evaluation.plot_maze import plot_multi_paths, PathFormat\n",
+                "from maze_transformer.evaluation.eval_model import decode_maze_tokens_to_coords, load_model_with_configs"
+            ]
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "metadata": {
+                "collapsed": false,
+                "pycharm": {
+                    "name": "#%%\n"
+                }
+            },
+            "outputs": [],
+            "source": [
+                "# Setup\n",
+                "device = configure_notebook(seed=42, dark_mode=True)\n",
+                "# We won't be training any models\n",
+                "torch.set_grad_enabled(False)\n",
+                "\n",
+                "# Get latest model\n",
+                "# this should point towards a directory containing a run. \n",
+                "# If you don't have any runs, you can create a dataset with `poetry run python scripts/create_dataset.py create ./data/maze 10 --grid_n=4`\n",
+                "# Then train a model with poetry run python scripts/train_model.py ./data/maze/g4-n10`\n",
+                "run_path = Path(\"../data/maze/g4-n10\")\n",
+                "assert run_path.exists(), f\"Run path {run_path.as_posix()} does not exist\"\n",
+                "model_path = list(sorted(run_path.glob(\"**/model.final.pt\"), key=os.path.getmtime))[\n",
+                "\t-1\n",
+                "].resolve()\n",
+                "model, cfg = load_model_with_configs(model_path)\n",
+                "maze_path = run_path / \"maze_tokens.jsonl\""
+            ]
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "metadata": {
+                "collapsed": false,
+                "pycharm": {
+                    "name": "#%%\n"
+                }
+            },
+            "outputs": [],
+            "source": [
+                "# generate a maze\n",
+                "grid_n: int = cfg.dataset_cfg.grid_n\n",
+                "maze: LatticeMaze = LatticeMazeGenerators.gen_dfs((grid_n, grid_n))\n",
+                "c_start = (0, 0)\n",
+                "c_end = (grid_n - 1, grid_n - 1)\n",
+                "\n",
+                "# solve the maze explicitly\n",
+                "path_true = np.array(maze.find_shortest_path(\n",
+                "\tc_start = c_start,\n",
+                "\tc_end = c_end,\n",
+                "))\n",
+                "\n",
+                "\n",
+                "# tokenize the maze\n",
+                "tokens = maze_to_tokens(maze, path_true, cfg.dataset_cfg.node_token_map)\n",
+                "path_start_index = tokens.index(SPECIAL_TOKENS[\"path_start\"])\n",
+                "maze_only_tokens = tokens[:path_start_index + 1]\n",
```

Comment:
It's interesting that we don't bother using the SolvedMaze here. It seems like at any point we have a maze and solution we could call `maze_to_tokens` or `solved_maze_to_tokens`. I think this is confusing.

Which made me think, maybe we should introduce a `MazeTokenizer` class which is responsible for tokenizing a maze with or without a soluition.....uh oh.

Maybe it would be sensible to create overloads of `maze_to_tokens` which take, respectively, [maze], [maze, solution], and [SolvedMaze]. But I don't think that is easy to do in python. So maybe we could use flags? [multipledispatch](https://stackoverflow.com/a/29091980/6631502) may help us here, but haven't looked in detail.

(Maybe this is more of a minor issue than I am making out)

Reply:
I think the problem here is that `solved_maze_to_tokens` shouldn't exist. It's really more of an internal method to `MazeDataset` to allow easier use of partials, but it's not really needed and I agree it causes confusion. I'll see if it's straightforward to remove as part of this PR.

I agree that a future enhancement could be to allow tokenizing mazes that don't have a solution, but I'm not sure about the best approach there and I think some research/thinking will be needed. It's not a very urgent use case (this is the only part of the codebase that we would use it currently), so I'd prefer to make a ticket for it and do it later.

Reply:
Okay, removed `solved_maze_to_tokens`.